### ISSUE TITLE
Weird difference in behaviour of 'method' function types vs regular function types

### ISSUE THREAD

**Issue by Torvin** [microsoft/TypeScript#40192]
I came across a weird difference in behaviour of 'method' function types vs regular function types. I thought doing

```ts
type Type = {
  fn(arg: X): Y
}
```
is exactly equivalent to 
```ts
type Type = {
  fn: (arg: X) => Y
}
```
but looks like this is not the case? Inferring whether `arg` is assignable works differently. See the code below.

**TypeScript Version:**  from at least 3.3 to 4.1.0-dev.20200822

**Search Terms:** arrow function type vs property function type

**Code**

```ts
type Fns = {
    propFn(arg: string | null): void
    arrowFn: (arg: string | null) => void
}

type X = {
    (arg: string | null): void
}

// notice the lack of `null` in argument's type
function testFn(arg: string) {
   console.log(arg.length) // possible runtime TypeError
}

const fns: Fns = {
    arrowFn: testFn, // type error as expected (ts2322 Type 'null' is not assignable to type 'string')
    propFn: testFn,  // somehow works?!
}
```

**Expected behavior:**
Assigning `testFn` to `propFn` should produce the same error `ts2322`

**Actual behavior:**
Assigning `testFn` to `propFn` succeeds, which looks wrong

**Playground Link:** [link](https://www.typescriptlang.org/play?ts=4.1.0-dev.20200822#code/C4TwDgpgBAYgdgZygXigbwFBW1MAnAezHgAoBDPAcwC4oFg8BLOSqAHyjgFcAbHgSloA3AowAmWHBUIB3eLXJVa9Ji3adeAlAD4oI8RgC+GDKEhQAGinSTsimnQbNWHbn0F7RE4xgD0vzgJgRgBjaGAAC2geMhCAaygCADMoAAM3HlSoZigKSi4AWwg4YAByJDMIDCSuOBDggjgoYAh6UjzlJxZ+GxwoEMaEAh4IADoeAkp7ceLKSJ7-XAIEBEYAIxGoPFrgoqgAFXAIAFE8QjwjEwHEYCgkxFp4JFRMPukCOThaFra4ABooItKlAIGcCHhckgIAAPSD1CBiKAkYAIABMAGZUaiDkcoKUMqVskg4EFIatKHAyBtwgRmrjyl1KKV+LZcIRiF9mq1gPAAYCAkMihEPlAZOC4ggAPwAQkuQA)
